### ISSUE TITLE
Missing pkg-config directory in path causes pkg-config lookup for Homebrew packages to fail from Homebrew's build environment

### ISSUE THREAD

**Issue by mologie** [Homebrew/homebrew#39384]
Hello,

this bug report is about inconsistencies with how Homebrew handles pkg-build files, which makes configure scripts that depend on pkg-config fail.

Specifically, I came across this issue while attempting to build a package that depends on libsodium and retrieves its paths using pkg-config (namely, php56-sodium). I was originally going to report this in homebrew-php's issue tracker, but this appears to be a more generic issue that I will explain now.

When attempting to locate libsodium using pkg-config from my shell, everything works as expected:

```
$ echo $HOMEBREW_PREFIX
/Users/oliver/.local

$ where pkg-config
/Users/oliver/.local/bin/pkg-config

$ pkg-config libsodium --libs
-L/Users/oliver/.local/Cellar/libsodium/1.0.2/lib -lsodium 
```

However, once I apply Homebrew's build environment, the query fails:

```
$ brew sh
Your shell has been configured to use Homebrew's build environment:
[...]

$ pkg-config libsodium --libs
Package libsodium was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsodium.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libsodium' found
```

I created a diff of the evironment and came to the conclusion that the following two environment variables are the cause of the error:

```
PKG_CONFIG_LIBDIR=/usr/lib/pkgconfig:/Users/oliver/.local/Library/ENV/pkgconfig/10.10
PKG_CONFIG_PATH=/Users/oliver/.local/opt/icu4c/lib/pkgconfig:/Users/oliver/.local/opt/libffi/lib/pkgconfig:/Users/oliver/.local/opt/openssl/lib/pkgconfig:/Users/oliver/.local/opt/sqlite/lib/pkgconfig:/Users/oliver/.local/opt/zlib/lib/pkgconfig
```

This is because Homebrew's generic pkgconfig directory is not referenced in PKG_CONFIG_PATH, but is implicitly used when the variable is absent. The libsodium pkg-config files are installed correctly:

```
$ find $HOMEBREW_PREFIX -name libsodium.pc
/Users/oliver/.local/Cellar/libsodium/1.0.2/lib/pkgconfig/libsodium.pc
/Users/oliver/.local/lib/pkgconfig/libsodium.pc
```

I noticed that for all packages which are explicitly listed in PKG_CONFIG_PATH, no entry in `$HOMEBREW_PREFIX/lib/pkgconfig` exists. Is this possibly related to them being marked as KEG-only? Here is an example

```
$ find $HOMEBREW_PREFIX -iname libffi.pc
/Users/oliver/.local/Cellar/libffi/3.0.13/lib/pkgconfig/libffi.pc
```

tl;dr: I believe that all that is required to fix the issue is to add Homebrew's default pkgconfig directory, `$HOMEBREW_PREFIX/lib/pkgconfig`, back to `PKG_CONFIG_PATH` when building Homebrew's environment. Without the directory being available in the path, Homebrew-installed libraries are invisible to pkg-config when used by the build process for other Homebrew packages. This causes the build process to fail for packages which always look up using pkg-config, regardless of the package already being available in the default paths.

**Comment by DomT4** [comment]
> back to PKG_CONFIG_PATH when building Homebrew's environment

I doubt this will happen. Too much chance of builds picking up random libraries and compiling silently against them. As far as I understand (which take with a pinch of salt, I'm often an idiot) `brew sh` shouldn't pick up any `pkgconfig` files outside of a build process when run under superenv. 

If `php56-sodium` adds a `pkg-config` dep it'll be able to access `libsodium`'s lib/pkgconfig directory and should build just fine. *(I believe)*.

**Comment by tdsmith** [comment]
> I noticed that for all packages which are explicitly listed in PKG_CONFIG_PATH, no entry in $HOMEBREW_PREFIX/lib/pkgconfig exists. Is this possibly related to them being marked as KEG-only?

Yes.

`brew sh` is probably not generally useful; the build environment inside a formula's `install` method will depend on the formula's dependencies, which will add additional paths to the `PKG_CONFIG_PATH`. You can drop a call to `interactive_shell` into the `install` method of the formula you're installing if you want to see what's actually happening.

If you're arguing that `brew sh` would be more useful if it set `PKG_CONFIG_PATH` differently, I think that might be reasonable. I'm a little surprised that it's necessary; the pkg-config man page says that the `PKG_CONFIG_PATH` is searched in addition to the default search path, but it seems to override it?

**Comment by mologie** [comment]
Thank you for your responses. I wrongly assumed that `brew sh` provides the same environment for all formulas, but it in fact does not. My error was to run php56-sodium's configure script from `brew sh` in an attempt to debug the issue that I experienced. The environment set by brew however provoked a different error which was reported using the same error message by the configure script. D'oh.

Everything is well now, the Homebrew's behaviour in `brew sh` appears to be intentional (which is fine), and the pkg-build path is set correctly during the build process which is visible using DomT4's interactive_shell trick.

@tdsmith:
> If you're arguing that brew sh would be more useful if it set PKG_CONFIG_PATH differently, I think that might be reasonable. I'm a little surprised that it's necessary; the pkg-config man page says that the PKG_CONFIG_PATH is searched in addition to the default search path, but it seems to override it?

The behaviour appears to be intentional as pointer out by DomT4. Digging further into the manual, I also find:
> PKG_CONFIG_LIBDIR - Replaces the default pkg-config search directory.

Which explains the behaviour that we observe: pkg-config does search the default directory, but brew happened to have replaced that one too. This is useful for avoiding conflicts with locally installed packages, but did indeed just bite me.

Because I now understand the purpose/concept of `brew sh`, I do not think that any changes to the command are required. An additional option which loads the environment of a specified formula (without the need of modifying that formula file to spawn an interactive shell) would certainly be helpful though.

**Comment by DomT4** [comment]
> I'm a little surprised that it's necessary

I thought `brew sh` was run under `superenv` so presumably we block finding the normal `HOMEBREW_PREFIX/lib/pkgconfig` directory? Could be wildly wrong though.

> set correctly during the build process which is visible using DomT4's interactive_shell trick.

Tim gets the credit for this suggestion I think :)

 If you can't build the formula without dropping into the interactive shell and adding `PKG_CONFIG_PATH` manually it's still worth kicking to the upstream repo and asking them to consider adding a `pkg-config` dep. After that dep is added, it *should* be able to find `libsodium` normally.
